rust/arrow/src/array/ffi.rs

```diff
@@ -0,0 +1,163 @@
+    dictionary: *mut FFI_ArrowSchema,
+    release: ::std::option::Option<unsafe extern "C" fn(arg1: *mut FFI_ArrowSchema)>,
+    private_data: *mut ::std::os::raw::c_void,
+}
+
+impl FFI_ArrowSchema {
+    /// allocates this struct. Unsafe as the consumer is responsible for owning it.
+    pub fn allocate() -> *mut Self {
+        Box::into_raw(Box::new(Self::new()))
+    }
+
+    fn new() -> Self {
+        Self {
+            format: std::ptr::null_mut(),
+            name: std::ptr::null_mut(),
+            metadata: std::ptr::null_mut(),
+            flags: 0,
+            n_children: 0,
+            children: std::ptr::null_mut(),
+            dictionary: std::ptr::null_mut(),
+            release: None,
+            private_data: std::ptr::null_mut(),
+        }
+    }
+}
+
+impl Drop for FFI_ArrowSchema {
+    fn drop(&mut self) {
+        match self.release {
+            None => (),
+            Some(release) => unsafe { release(self) },
+        };
+    }
+}
+
+/// ABI-compatible struct for ArrowArray from C Data Interface
+#[repr(C)]
+#[derive(Debug, Clone)]
+pub struct FFI_ArrowArray {
+    length: i64,
+    null_count: i64,
+    offset: i64,
+    n_buffers: i64,
+    n_children: i64,
+    buffers: *mut *const ::std::os::raw::c_void,
+    children: *mut *mut FFI_ArrowArray,
+    dictionary: *mut FFI_ArrowArray,
+    release: ::std::option::Option<unsafe extern "C" fn(arg1: *mut FFI_ArrowArray)>,
+    private_data: *mut ::std::os::raw::c_void,
+}
+
+impl FFI_ArrowArray {
+    /// allocates this struct. Unsafe as the consumer is responsible for owning it.
+    pub fn allocate() -> *mut Self {
+        Box::into_raw(Box::new(Self::new()))
+    }
+
+    fn new() -> Self {
+        Self {
+            length: 0,
+            null_count: 0,
+            offset: 0,
+            n_buffers: 0,
+            n_children: 0,
+            buffers: std::ptr::null_mut(),
+            children: std::ptr::null_mut(),
+            dictionary: std::ptr::null_mut(),
+            release: None,
+            private_data: std::ptr::null_mut(),
+        }
+    }
+}
+
+impl Drop for FFI_ArrowArray {
+    fn drop(&mut self) {
+        match self.release {
+            None => (),
+            Some(release) => unsafe { release(self) },
+        };
+    }
+}
+
+/// Struct containing the necessary data and schema to move an Array from and to the C Data Interface.
+#[derive(Debug, Clone)]
+pub struct ArrowArray {
+    array: Arc<FFI_ArrowArray>,
+    schema: Arc<FFI_ArrowSchema>,
+}
+
+impl ArrowArray {
+    /// constructs an `ArrowArray` from two raw pointers.
+    pub fn allocate() -> (*mut FFI_ArrowArray, *mut FFI_ArrowSchema) {
+        (FFI_ArrowArray::allocate(), FFI_ArrowSchema::allocate())
+    }
+
+    pub unsafe fn from_raw_pointers(
+        array: *mut FFI_ArrowArray,
+        schema: *mut FFI_ArrowSchema,
+    ) -> Self {
+        ArrowArray {
+            array: Arc::from_raw(array),
+            schema: Arc::from_raw(schema),
```

Comment:
Please excuse my ignorance, but according to https://doc.rust-lang.org/beta/std/sync/struct.Arc.html#method.from_raw , this only works if the pointers come from `to_raw_pointers`?

Reply:
Good catch @pitrou. It was a bug: I wanted to use only `Arc`, no `Box` (as the rest of the code is like that).